### ISSUE TITLE
[FR] subplotgrid function

### ISSUE THREAD

**Issue by Datseris** [JuliaPlots/Makie.jl#1581]
Would be nice to have something equivalent with PyPlot's `subplots` function:

```
fig, axs = subplotgrid((X, Y); sharex, sharey)
```
such a function would make an XxY grid layout, and optionaly share their x and/or y axis (also hiding ticks if axis are shared).

(I'll do this, assign me on GitHub please)

**Comment by Datseris** [comment]
@SimonDanisch I'm not making a PR yet because I just don't know where to put this function. But here is the code:
```julia
using GLMakie

"""
    subplotgrid(m, n; sharex = false, sharey = false, kwargs...) -> fig, axs
Create a grid of `m×n` subplots in a new figure and return the figure and the 
matrix of axis. Optionally make every row share the y axis, and/or every column 
to share the x axis. Any additional keywords are given to `Figure` generation.
"""
function subplotgrid(m, n; sharex = false, sharey = false, kwargs...)
    fig = Figure(;kwargs...)
    axs = Matrix{Axis}(undef, m, n)
    for i in 1:m
        for j in 1:n
            axs[i,j] = Axis(fig[i,j])
        end
    end
    if sharex
        for i in 1:m-1
            for j in 1:n
                linkxaxes!(axs[i,j], axs[m,j])
                hidexdecorations!(axs[i,j]; grid=false)
            end
        end
    end
    if sharey
        for j in 2:n
            for i in 1:m
                linkyaxes!(axs[i,j], axs[i,1])
                hideydecorations!(axs[i,j]; grid=false)
            end
        end
    end
    return fig, axs
end

fig, axs = subplotgrid(3,2; sharex = true, sharey = false)
display(fig)
```
 feel free to give a review and/or tell me where to put it.

**Comment by Datseris** [comment]
![image](https://user-images.githubusercontent.com/19669089/149034180-b41d9161-e739-4277-81c1-e1941aa9061b.png)

**Comment by Datseris** [comment]
bump!

**Comment by Datseris** [comment]
I will also make an in-place version that takes an existing figure or `GridLayout` as an argument and adds the axis. Just tell me where to actually put it to make a PR :D

**Comment by jkrumbiegel** [comment]
Maybe this would be good for a `Block`, it could be called FacetGrid for example. That's because it could benefit from spanning labels on the axes and maybe boxes for the titles such as commonly found in ggplot.

**Comment by Datseris** [comment]
I am not aware of what a `Block` is (https://makie.juliaplots.org/stable/search/index.html?q=Block ). Is it internal?

**Comment by jkrumbiegel** [comment]
It's the new implementation of what Layoutable was, on master. A building block possible made of other building blocks, that you can place in a grid as one rectangular object.

**Comment by Datseris** [comment]
Yes, I'll make it a block once I have the instructions of how to do so. Is it just a simple wrap of a `GridLayout`? In reality, the function I've coded above can just return a `GridLayout`.

**Comment by jkrumbiegel** [comment]
There's an initial implementation here https://github.com/JuliaPlots/Makie.jl/pull/1827

```julia
begin
    f = Figure()
    fg = MakieLayout.FacetGrid(f[1, 1], 3, 4, sharey = :all,
        rowlabels = ["Hi", "Whats", "Up"])
    fg.columnlabels = ["a", "b", "c", "d"]
    lines!(fg[1, 2], cumsum(randn(50)))
    lines!(fg[3, 4], cumsum(randn(50)))
    Label(f[0, :], "A FacetGrid demonstration", tellwidth = false)
    f
end
```

![grafik](https://user-images.githubusercontent.com/22495855/163848824-6016267b-dc26-4335-8946-0c58f7880494.png)

I'm not sure yet if I like making an object out of this or if it should just make axes wherever. The options to style this are potentially infinite, and it's not going to be practical to implement them all with overlapping keywords etc. For example position of label boxes, if it's a full grid or if there are positions missing, etc. However, using a Block does make the thing somewhat nicely self contained.

**Comment by Datseris** [comment]
I really hope the Makie developers strongly consider adding the above `subplotgrid` function to the API and exporting it, because it has become rather tedius to copy paste it in 20 different science projects and then include the file with this definition in every one scripts in these projects. Given the prevalence of "facet/grid plots" in science, this function is likely to be widespread used if available.

**Comment by jkrumbiegel** [comment]
Yes, definitely. I need a bit more time to think about how exactly to integrate it, as this function is only one example of a multi-axis (or object) plotting recipe, for which we need better infrastructure anyway. So I don't want to include this as is only to come up with a unified way a bit later, leading to more breaking changes.

**Comment by Datseris** [comment]
I've updated my [original post](https://github.com/MakieOrg/Makie.jl/issues/1581#issuecomment-1010439260) with an even better and more general version of the code, that I Am particularly prooud about! I use this code in literally every science project I have!

**Comment by jkrumbiegel** [comment]
Yeah I agree, let me try to prepare a PR soon, I have some small ideas about the API that I'd want to try. But it would be very useful to have
